### PR TITLE
EventType autocreate

### LINKED ISSUE #3174: Implement the auto-create feature flag for the Knative broker implementation for apache Kafka
Similar to https://github.com/knative/eventing/issues/6909, we should add support for this new feature in the Knative broker implementation for apache Kafka

### PR DESCRIPTION
**Pull request by Cali0707**
Fixes #3174 #3181 

## Proposed Changes

- Create an EventTypeCreator class to handle the creation of eventtypes
- Call the create method in the receiver ingress on successfully receiving an event

These changes work as follows:
```bash
kn broker create my-broker --class Kafka
Broker 'my-broker' successfully created in namespace 'default'.

 kn-event-linux-amd64 send --to Broker:eventing.knative.dev/v1:my-broker --type type.eventing.example.1
Jan 10 12:35:42.087   INFO Event (ID: e0a551a7-7f1e-482e-aaa0-10fc5c076326) have been sent.

kubectl get eventtypes.eventing.knative.dev -A
No resources found

kubectl apply -f enable-autocreate.yaml                           
configmap/config-features configured

kn-event-linux-amd64 send --to Broker:eventing.knative.dev/v1:my-broker --type type.eventing.example.1
Jan 10 12:52:07.555   INFO Event (ID: a8c83ebf-d8a6-4fb0-9fb1-c2d67f8a21e9) have been sent.

kubectl get eventtypes.eventing.knative.dev -A                                                        
NAMESPACE   NAME                                            TYPE                      SOURCE             SCHEMA   REFERENCE NAME   REFERENCE KIND   DESCRIPTION                             READY   REASON
default     et-my-broker-cc78b6ed9ee7697817cf61de134725aa   type.eventing.example.1   kn-event/v1.10.0            my-broker        Broker           Event Type auto-created by controller   True    

```

**Release Note**

```release-note
:gift: EventType Autocreate now works for the Eventing Kafka Broker and Eventing Kafka Channel
```

**Docs**

### CODE REVIEW

**Comment by Leo6Leo** [comment]
@Cali0707  Tested in my local environment, the event type works perfectly! Great work!!!

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            this.messageDigest = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException ignored) {
+            this.messageDigest = null;
+        }
+    }
+
+    private String getName(CloudEvent event, DataPlaneContract.Reference reference) {
+        final var suffixString = event.getType() + event.getSource() + reference.getNamespace() + reference.getName();
+        this.messageDigest.reset();
+        this.messageDigest.update(suffixString.getBytes());
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:50]: what happens when this is called concurrently ? I guess this is not thread safe ?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        final var suffix = Hex.encodeHexString(this.messageDigest.digest());
+        final var name = String.format("et-%s-%s", reference.getName(), suffix).toLowerCase();
+        if (name.length() > DNS1123_SUBDOMAIN_MAX_LENGTH) {
+            return name.substring(0, DNS1123_SUBDOMAIN_MAX_LENGTH);
+        }
+        return name;
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:56]: Is this the same algorithm we use in Eventing core ?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    private boolean eventTypeExists(String etName, DataPlaneContract.Reference reference) {
+        var et = this.eventTypeClient
+                .inNamespace(reference.getNamespace())
+                .withName(etName)
+                .get();
+        return et != null;
+    }
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:65]: Are we using an actual client ? should we use a lister here ?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                        .withUid(ownerReference.getUuid())
+                        .build())
+                .withNamespace(ownerReference.getNamespace())
+                .withName(name)
+                .withType(event.getType())
+                .withSource(event.getSource())
+                .withSchema(event.getDataSchema())
+                .withDescription("Event Type auto-created by controller");
+
+        this.eventTypeClient.resource(et.build()).create();
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:93]: we should avoid failing when the resource exist (like when it is created by some other process between checking exists and this line)

**Review by pierDipi — COMMENTED** [review]
Are you also planning to add E2E tests ?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java (hunk 1 above)
Comment (Cali0707) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:50]: Yeah I doubt this is thread safe, but currently it is just used in the `IngressRequestHandlerImpl` which I thought was 1 per verticle (so no concurrency worries?), should I change this to be thread safe?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java (hunk 2 above)
Comment (Cali0707) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:56]: I think so, I tried to copy from https://github.com/knative/eventing/blob/42af160c3cfc244e26fa7bcd5db2c670ebdcf326/pkg/eventtype/eventtypes.go#L48-L52

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java (hunk 3 above)
Comment (Cali0707) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:65]: This is an actual client as I wasn't sure how to make a lister to our custom class - I'll try and update it to be a lister

**Comment by Cali0707** [comment]
For E2E tests, I was planning on porting those from eventing core: https://github.com/knative/eventing/blob/42af160c3cfc244e26fa7bcd5db2c670ebdcf326/test/experimental/eventtype_autocreate_test.go#L32-L74

But, I'm not sure if we have a job for experimental tests so I think I need to set that up first

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java (hunk 4 above)
Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:93]: We need to be sure on what happens when already exists and handle that case accordingly

Comment (Cali0707) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:93]: I looked on their github and they said that the methods generally throw exceptions, so I'll add a try-catch to handle that case

**Comment by Cali0707** [comment]
/cc @pierDipi 

I've added e2e tests to this, and addressed your comments (I think)

**Comment by dsimansk** [comment]
@Cali0707 there is an errror during code compile phase. Have you committed all the changes perhaps?

**Comment by Cali0707** [comment]
> @Cali0707 there is an errror during code compile phase. Have you committed all the changes perhaps?

Looks like I had some uncommited changes locally, just pushed. Thanks for catchig @dsimansk

File: control-plane/config/eventing-kafka-broker/200-controller/200-controller-cluster-role.yaml [hunk 5]
```diff
@@ -213,6 +213,17 @@ rules:
     verbs:
       - update
 
+  # resources needed to grant eventtype autocreate rbac to namespaced data plane component
+  - apiGroups:
+      - "eventing.knative.dev"
+    resources:
+      - "eventtypes"
```
Comment (dsimansk) [inline control-plane/config/eventing-kafka-broker/200-controller/200-controller-cluster-role.yaml:220]: Could be merged into L190 group?

**Comment by Cali0707** [comment]
Looks like the lister isn't working/updating withe new values. Will need to look into this more

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Comment by matzew** [comment]
/retest

are we sure these no flakes, @Cali0707

**Comment by Cali0707** [comment]
Not sure, I wasn't seeing them before but some other stuff has been merged since then. Let me take a look today @matzew

**Comment by Cali0707** [comment]
@matzew it looks like the unit test sometimes didn't wait long enough for the lister to update, I increased the timeout a bit to fix

**Comment by matzew** [comment]
@Cali0707 needs a rebase, unfortunately :see_no_evil:

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java (hunk 1 above)
Comment (pierDipi) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:66]: Yes, I would opt to being safer here

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            Lister<EventType> eventTypeLister,
+            Vertx vertx) {
+        this.eventTypeClient = eventTypeClient;
+        this.eventTypeLister = eventTypeLister;
+        this.executor = vertx.createSharedWorkerExecutor("et-creator-worker", 1);
+        try {
+            this.messageDigest = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException ignored) {
+            this.messageDigest = null;
+        }
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:60]: Catching and leaving it to null, will just delay the exception to the runtime, can we just not catch it or if it's unchecked exception, we could wrap it into a runtime exception?

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            return name.substring(0, DNS1123_SUBDOMAIN_MAX_LENGTH);
+        }
+        return name;
+    }
+
+    private EventType eventTypeExists(String etName, DataPlaneContract.Reference reference) {
+        if (this.eventTypeLister == null) {
+            logger.debug("no eventtype lister, creating eventtype and might run into failure");
+            return null;
+        }
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:79]: Should we fail fast here in the constructor?

File: data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImplTest.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+                    try {
+                        // make sure that the informer has time to resync, the webhook doens't seem to run in the tests
+                        Thread.sleep(2000);
+                    } catch (InterruptedException e) {
+                        vertxTestContext.failNow(e);
+                    }
+                    return eventTypeCreator
+                            .create(event, reference)
+                            .onFailure((exception) -> {
+                                logger.warn("failure occured, closing informer", exception);
```
[the remark this answers is not in the document]
Comment (Cali0707) [reply data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImplTest.java:155]: Good catch, thanks!

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java (hunk 6 above)
Comment (Cali0707) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/eventtype/EventTypeCreatorImpl.java:60]: Currently it doesn't throw a runtime exception, the autocreate just becomes a no-op (no eventtype gets created). Do you think it's better to fail here instead?

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

**Pull request merged**